### DISCUSSION TITLE
How to filter output data in the proper format to be able to create variables that I need in the next part of the task?

### DISCUSSION THREAD

**Discussion by Sivileo** [jspsych/jsPsych#1751] [Q&A] [answered]
Hi everyone,

I´m working on a new experiment where on its first part, 
participants rate a set of images (0-100).
 
For the second part of the experiment, I want to select a subset of images
depending on these ratings. Specifically, I would like to create an array
that contains image objects to use as stimuli in a task context,
e.g, low-score images vs. high-score images.

I expect to find a way to obtain an array with the selected subset of images
(somehow using the output data) and my first attempt has been using jspsych.data.get().filter().filtercostum(), 
but using this approach I instead obtain an object that I don´t know how to manage.

Is there a way to obtain the filtered collection of images as I was expecting,
or should use a different approach?.

Also, I am wondering, how could I display the filtered output data? 
(It would help me to check the "var" information that I am creating to see if it´s what I am expecting)

Thank you very much for your help in advance,
 Silvia

**Comment by becky-gilbert** [comment, marked as answer]
Hi @Sivileo, this is a little hard to explain without some example code to work with, so feel free to post a minimal example if you have one.

I think you're on the right track with the approach you suggested. My general approach would be to:
1. Filter the jsPsych data after the first part and before the second part, based on some unique properties in the data that select for the ratings trials. Usually the easiest way to do this is to put a flag in the data for the trial so that you can filter based on that, e.g.
```js
var rating_trial = {
  // other trial parameters...
  data: {task_part: 'rating'}
};

// then, after part 1, you can filter the data to get all of the image ratings:
var ratings_trials = jsPsych.data.get().filter({task_part: 'rating'});
```
2. Once you have all of the image ratings, categorize them as high or low. (There are lots of ways to do this, so it would be helpful to know more about how you want to handle this. For instance, you could use pre-defined cut-off values, but that only works if you know that people are using the whole scale, or if you don't mind if they don't have any trials that fall in your high and/or low categories. Another option is to order the ratings from lowest to highest, and take the first X values for the low category and last X trials for the high category. In this case you might want to also include a check to see if the high and low values are actually different. For instance, if someone rates all of the images with 0 or 100, then the high and low categories would have the same ratings.)
Anyway, here's one way to do this:
```js
// filter the ratings trials again to see if the response meets the criteria
// for the low/high category
var high_rated_images = ratings_trials.filterCustom(function(trial) {
  return trial.response > 75;
});
var low_rated_images = ratings_trials.filterCustom(function(trial) {
  return trial.response < 25;
});
// note that you can also do this in a single step, by chaining togther the filter and filterCustom functions
// as you suggested in your post:
// var high_rated_images = jspsych.data.get().filter({task_part: 'rating'}).filterCustom(function(trial) {
//   return trial.response > 75;
// });
// var low_rated_images = jspsych.data.get().filter({task_part: 'rating'}).filterCustom(function(trial) {
//   return trial.response < 25;
// });

// check that the low and high arrays contain all of the images that meet your criteria
console.log('high rated images: ', high_rated_images); 
console.log('low rated images: ', low_rated_images);
```
3. For your second part, you can run a set of trials for each of the images in the `low_rated_images` and `high_rated_images` arrays. To do this, I would use a combination of conditional timelines and looping timelines. The looping timeline is needed to loop over the images in the two arrays. The conditional timeline is needed to check if either of the image arrays is empty - if so, then skip the looping timeline (because there's nothing to loop over), otherwise run the looping timeline. Here's what it might look like:
```js
// create a global variable to store the current image stimulus
// the value of this variable will be changed in the looping function
var current_image;
// create a global variable to track the index of the current image in the low/high arrays
var count = 0;

var repeat_image_trial = {
  stimulus: function() {
    // use a dynamic parameter to get the current value of the current_image variable
    return current_image;
  },
  // other parameters for the trial that presents the low/high-rated images...
};

var low_rated_images_loop = {
  timeline: [repeat_image_trial],
  looping_function: function() {
    count++; 
    if (count < low_rated_images.length) { 
      // after increasing the trial count, if there are still images in the array, then set current_image
      current_image = low_rated_images[count];
      return true;
    } else {
      // if the trial count has exceeded the number of images in the array, then we're done looping
      return false;
    }
  }
};

var low_rated_images_conditional = {
  timeline: [low_rated_images_loop]
  conditional_function: function() {
    if (low_rated_images.length > 0) {
      // if there are any images in the low_rated_images array, then run the low_rated_images_loop procedure
      return true;
    } else {
      // otherwise skip the low_rated_images_loop procedure
      return false;
    }
  }
};
// add the conditional timeline to your main experiment timeline
timeline.push(low_rated_images_conditional);
```

> Also, I am wondering, how could I display the filtered output data?
> (It would help me to check the "var" information that I am creating to see if it´s what I am expecting)

Yes, you can include the function `console.log()` to print out text and/or the current value of any of your variables. To view the JavaScript console, open the browser's developer tools (usually F12, or Ctrl+Shift+i, or menu -> More tools -> Developer tools). This is a great way to debug problems!

**Reply by Sivileo** [reply]
Hi, @becky-gilbert  thanks for so quick response,
and sorry about my generalistic question.
I´m a beginner in programming and I don´t have too much experience.
So let´s start again using some examples.

For the first part, I´m creating an array that looks like this:

```
var all_img =[
{stim_img: '/0003.png'},
{stim_img: '/0008.png'},
{stim_img: '/0009.png'},
{stim_img: '/0011.png'},
{stim_img: '/0024.png'},
{stim_img: '/0028.png'},....]
```

I´m using the timeline variable because I need to use these images as
a random stimulus for each participant.
Also, I will ask them to rate the images based on two criteria (T and H).
So, I´m putting a flag to each rating criteria 
(as you comment on the first point).
e.g. T_rating and H_rating.

This is an example for the T_rating trial that I´m using:

```
var T_rating = {
   type:'image-slider-response',
   stimulus: jsPsych.timelineVariable('stim_img'),
   stimulus_height: 300,
   maintain_aspect_ratio: true,
   labels: ['Most UnT', 'Most T'],
   slider_start: function() {
         return jsPsych.randomization.sampleWithoutReplacement([...], 1)
        },
   require_movement: true,
   data:{ 
    stage: 'TRatings'
   }
  };
```

```
var procedure_T = {
  timeline: [T_rating],
  timeline_variables: all_img,
  randomize_order: true
 };

timeline.push(procedure_T);
```

The first part work apparently well. I´m checking this using: 

> jsPsych.data.displayData('csv').

But now, when I apply the filter (a simple one as you commented to me):

```
var T_Rating_trials = jsPsych.data.get().filter({stage: 'TRatings'});

console.log('T_Rating_trials:', T_Rating_trials );
```
On the log of the console seems that I´m creating an empty object.
Inside this object, I can see many categories that are equal for both criteria.
(This also happen when I use both filters together)

```
T_Rating_trial: Object
> addToAll: ƒ (properties)
> addToLast: ƒ (properties)
> count: ƒ ()
> csv: ƒ ()
......
```
I get the feeling that I´m missing something really important. If I
can´t apply this filter properly,
then I can´t still apply the other points that you said to me.

Again thank you very much in advance and also thank you for your patience.
Silvia

**Reply by becky-gilbert** [reply]
Ahh, sorry, that was my mistake! The `filterCustom` function will return a [jsPsych data collection](https://www.jspsych.org/core_library/jspsych-data/#datacollection), which is a special kind of object that has lots of built in methods for doing things that you might want to do to a set of trials (e.g. count the number of trials, add properties to all of the trials, convert it to CSV format, etc.). One of the methods available is called [`values`](https://www.jspsych.org/core_library/jspsych-data/#values), and it gives you the raw data values for the data collection, which is what we want to do in this case. So we just need to use the `values` function on the filtered data when creating the `high_rated_images` and `low_rated_images` arrays.
```js
var high_rated_images = ratings_trials.filterCustom(function(trial) {
  return trial.response > 75;
}).values();
var low_rated_images = ratings_trials.filterCustom(function(trial) {
  return trial.response < 25;
}).values();
```
You could also use the `values` function to check the data in `T_Rating_trials`. But keep in mind that you don't want to convert the `T_Rating_trials` data collection into the raw data values if you need to filter it again (e.g. using `filterCustom` to get the high/low rated images, as above). So if that's the case, then just use the `values` function when you want to look at the data with `console.log`, not when you create the variable:
```js
// leave this as a data collection so that you can use the filterCustom function on it 
var T_Rating_trials = jsPsych.data.get().filter({stage: 'TRatings'});
// use the values function here to get the raw values from the data collection, and print that to the console
console.log('T_Rating_trials:', T_Rating_trials.values() );
```
I hope that makes sense?

**Reply by Sivileo** [reply]
Hi, @becky-gilbert 

that makes more sense,
but the problem persists.

Just applying the basic filter:

```
var T_Rating_trials = jsPsych.data.get().filter({stage: 'TRatings'});
console.log('T_Rating_trials:', T_Rating_trials.values() )
```

On the log of the console, now seems 
that I´m creating an empty array instead of an empty object.

I have tried this in some examples that you include in the jspsych library 
and I have obtained the same result (empty arrays).

I'm sure I'm failing. I wonder if you could give me some very simple example that works
and I can apply, and compare it. Perhaps that way I will discover what I am doing wrong.
I suppose it has to be something prior to the filter.

Thank you so much for everything

**Reply by becky-gilbert** [reply]
Hi @Sivileo, hmm, that code looks fine to me... 
Can I see where in your script you added that code to get the T_Rating trials? I just want to check that the code is running _during_ the experiment, after the data from those trials has been collected.

**Reply by Sivileo** [reply]
Hi @ becky-gilbert,

Yesterday, in the meeting with my project manager, he told me that he does not see it as necessary. At the moment, we are going to try to solve it together. He will help me to review the code in depth to see if it is my fault (I am sure it is). If we couldn't find the solution, there would be no problem sending you the code. But he thinks we have to keep working on it (I agree).

As all this is new to me, I am sure that the overwhelm and frustration is making me make the most stupid of mistakes. Meanwhile, I will review the youtube tutorials (again and again) and review the documentation on the web. Everything I need is there (that's how I got to do the first part).

Finally, I want to tell you that thanks for all the help. I think your team has done a great job, and that your YouTube and web tutorials are very useful.

Thanks for everything again.
Silvia

**Reply by becky-gilbert** [reply]
Hi @Sivileo, no problem at all. Most researchers have no experience with web programming, and everyone makes mistakes while learning. We'll try to help if we can, just let us know!

**Reply by moramaldonado** [reply]
Hi @becky-gilbert, 
I was looking for a solution to a similar problem and I found your answer to this threat super helpful. There is only one part of it which doesn't work for me, and my guess is that I am not creating the array in the right moment during the experiment. I don't know how to solve it, so I thought maybe you have an idea. 

In my case, I am  doing a recursive sorting task. I am recovering the last 4 trials from the first part of the experiment and using the images sorted in piles in those 4 trials as input for the new sorting. The experiment fails when I dynamically call the stimuli (but the console.log  correctly shows me the array every time I call it).

Thank you very much in advance!

Here is the code:

This is the moment where I create the array I will use later on.

`   var pile;
    var count = 0;
    var piles = [];
`

  `  var part1 = {
        timeline: [{...}],
        on_timeline_finish: function() {
            var data_part2= jsPsych.data.get().last(4).values()
            for(let i=0; i< data_part2.length; i++) {
                if (data_part2[i].pile1.length > 1) {
                    piles.push(data_part2[i].pile1)
                }
                if (data_part2[i].data_part2.length > 1) {
                    piles.push(data_part2[i].pile2)
                }
            }
            return piles
        }
    }`

This is the loop function:

`  var sort_loop = {
        timeline: [{
            stimuli: function() {
                return pile
            }}],
        loop_function: function() {
            console.log(piles);
            count++;
            if (count < piles.length) {
                pile = piles[count];
                console.log(pile)
                return true;
            } else {
                return false;
            }
        }
    };`

This is the conditional: 

`    var sort_conditional = {
        timeline: [sort_loop],
        conditional_function: function() {
            console.log(piles)
            if (piles.length > 0) {
                return true;
            } else {
                return false;
            }
        }
    };`

**Reply by becky-gilbert** [reply]
Hi @moramaldonado, I think the problem might be with using the `pile` variable as the stimulus in the `sort_loop` timeline. It looks like this variable is defined in the loop function, but looping timelines execute _before_ the loop function is called. So I'm wondering if the `pile` variable isn't defined when the loop timeline runs for the first time. One way to check is to add `console.log(pile)` to your stimulus function, before the return statement.

If that is the problem then you just need to initialize the `pile` variable with `pile = piles[count]` sometime after `piles` is created but before the experiment reaches the looping timeline, for instance in the `on_timeline_finish` function or in the conditional function for `sort_conditional`. Does that work?

**Reply by moramaldonado** [reply]
Hi Becky, 
Just to signal that I have done as you suggested: I define the variable pile inside the conditional function and it works perfectly. 

Thanks a lot!

**Reply by becky-gilbert** [reply]
Great, thanks for letting us know!
